TestGracefulStop flake

Happened in 1.9 + 386 and 64-bit.

```
--- FAIL: TestGracefulStop (10.01s)
	leakcheck.go:109: Leaked goroutine: goroutine 63992 [chan receive]:
		google.golang.org/grpc/test.(*delayConn).Read(0x19b7ee40, 0x19af0000, 0x8000, 0x8000, 0x0, 0x0, 0x2)
			/home/travis/gopath/src/google.golang.org/grpc/test/gracefulstop_test.go:102 +0x2c
		bufio.(*Reader).Read(0x19bc4450, 0x19772320, 0x9, 0x9, 0x8072f24, 0x19a7a208, 0x2)
			/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/bufio/bufio.go:213 +0x22b
		io.ReadAtLeast(0x86b3d90, 0x19bc4450, 0x19772320, 0x9, 0x9, 0x9, 0x0, 0x1971d140, 0x804c394)
			/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/io/io.go:309 +0x68
		io.ReadFull(0x86b3d90, 0x19bc4450, 0x19772320, 0x9, 0x9, 0x19a7fecc, 0x8096b70, 0x19c260e0)
			/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/io/io.go:327 +0x45
		golang.org/x/net/http2.readFrameHeader(0x19772320, 0x9, 0x9, 0x86b3d90, 0x19bc4450, 0x0, 0x0, 0x0, 0x0, 0x0)
			/home/travis/gopath/src/golang.org/x/net/http2/frame.go:237 +0x66
		golang.org/x/net/http2.(*Framer).ReadFrame(0x19772300, 0x197f01c8, 0x1971d100, 0x86b7b60, 0x19b54070)
			/home/travis/gopath/src/golang.org/x/net/http2/frame.go:492 +0x82
		google.golang.org/grpc/transport.(*http2Client).reader(0x1976a410)
			/home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:1154 +0x2b
		created by google.golang.org/grpc/transport.newHTTP2Client
			/home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:273 +0xaaf
FAIL
```

I've also encountered the following error:

```
--- FAIL: TestGracefulStop (0.00s)
	gracefulstop_test.go:209: EmptyCall() = rpc error: code = Unavailable desc = transport is closing; want <nil>
FAIL
```

This looks like a production code problem vs. a test problem.

--- a/test/gracefulstop_test.go
+++ b/test/gracefulstop_test.go
@@ -38,6 +38,7 @@ type delayListener struct {
 	acceptCalled chan struct{}
 	allowCloseCh chan struct{}
 	cc           *delayConn
+	dialed       bool
 }
 
 func (d *delayListener) Accept() (net.Conn, error) {
@@ -70,6 +71,13 @@ func (d *delayListener) allowClientRead() {
 }
 
 func (d *delayListener) Dial(to time.Duration) (net.Conn, error) {
+	if d.dialed {
+		// Only hand out one connection (net.Dial can return more even after the
+		// listener is closed).  This is not thread-safe, but Dial should never be
+		// called concurrently in this environment.
+		return nil, fmt.Errorf("no more conns")
+	}
+	d.dialed = true
 	c, err := net.DialTimeout("tcp", d.Listener.Addr().String(), to)
 	if err != nil {
 		return nil, err

